ADBDEV-4726-69 Add null-check for first_rule

dnskvlnk opened the pull request:
Add null-check for first_rule

write_out_rule dereferences first_rule before null-check making segfault
possible. This patch adds an assertion

Commit:
Add null-check for first_rule

write_out_rule dereferences first_rule before null-check making segfault
possible. This patch adds null-check

RekGRpth commented on src/backend/utils/adt/ruleutils.c:
```diff
@@ -10867,7 +10867,7 @@ write_out_rule(PartitionRule *rule, PartitionNode *pn, Node *start,
 	char *str;
 
-	if (!*first_rule)
+	if (first_rule && !*first_rule)
```
I think the error is [here](https://github.com/arenadata/gpdb/pull/765/files#diff-6c12827097e95912476267d73f3f56e8250b13aad37ab4286c3c7fdab7f30ec7R10896) and should be like this:
```c
		if (strlen(body->buf->data) && !*first_rule &&
```

dnskvlnk replied:
Take a look, `first_rule` isn't dereferenced here, you changed the line in your example
https://github.com/arenadata/gpdb/blob/604bb5312052f3a1d8b62eb38ff3a51231fcd91d/src/backend/utils/adt/ruleutils.c#L10896

RekGRpth replied:
> Take a look, `first_rule` isn't dereferenced here, you changed the line in your example

Yes, I think first_rule should dereferenced here

dnskvlnk replied:
Then it would be easier just to assert that `first_rule` is not null

RekGRpth replied:
I think there is a semantic error here: we don't want to print a space before the first rule.
https://github.com/arenadata/gpdb/blob/604bb5312052f3a1d8b62eb38ff3a51231fcd91d/src/backend/utils/adt/ruleutils.c#L10896-L10898

dnskvlnk replied:
Added a condition

RekGRpth replied:
first_rule should not be NULL

RekGRpth replied:
write_out_rule is a static function and it is always called with not-NULL first_rule pointer

dnskvlnk replied:
Added assertion and removed checks

Commit:
Fix type

Commit:
Add null-check for first_rule

write_out_rule dereferences first_rule before null-check making segfault
possible. This patch adds an assertion
